Add ability to add back up option for form linking

mkangia opened the pull request:
[doc](https://trello.com/c/wekRPsTb/137-app-builder-changes-for-enikshay-phase-2b?menu=filter)

mkangia committed:
Add ability to add back up option for form linking

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/form_workflow.js:
```diff
@@ -74,6 +75,24 @@ hqDefine('app_manager/js/form_workflow.js', function() {
 
+    FormWorkflow.prototype.workflowBackUpOptions = function() {
+        // allow all options as backups except the one for form linking
+        var backup_options = _.omit(this.labels, function(key, value) { return value == FormWorkflow.Values.FORM })
```
Expected '===' and instead saw '=='. (eqeqeq)
Missing semicolon. (semi)

mkangia committed:
dimagimon complaints

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/form_workflow.js:
```diff
@@ -74,6 +75,24 @@ hqDefine('app_manager/js/form_workflow.js', function() {
 
+    FormWorkflow.prototype.workflowBackUpOptions = function() {
+        // allow all options as backups except the one for form linking
+        var backup_options = _.omit(this.labels, function(key, value) { return value === FormWorkflow.Values.FORM })
```
Missing semicolon. (semi)

mkangia committed:
better alignment in the view

snopoke commented on corehq/apps/app_manager/suite_xml/post_process/workflow.py:
```diff
@@ -306,6 +306,29 @@ def frame_children_for_module(module_, include_user_selections=True):
+                        # or a non-autoselect datum
+                        last = frame_children.pop()
+
+                    stack_frames.append(StackFrameMeta(None, frame_children))
```
Instead of just copying this from above can you not extract it to a function?

For this to work correctly as a fall-back option you'll need to set the `if_clause` in the `StackFrameMeta` to be the inverse of all the `link.xpath` expressions.

mkangia replied:
> Instead of just copying this from above can you not extract it to a function?

Absolutely will do.

@sheelio said we can just skip any if_clause in this.
This would be the fall back in case all conditions fail but yes would be last path the app will take any way after the form is filled up. May be changing the name to something like "last view" would make this clear.

mkangia replied:
Ended up doing what @snopoke suggested.

snopoke commented on corehq/apps/app_manager/templates/app_manager/v1/partials/form_workflow.html:
```diff
@@ -43,6 +43,18 @@
+                        value: workflowbackup,
+                        optionsCaption: 'Choose a backup...'
+                    "></select>
+                </div>
```
This should only be visible if form linking is selected.

mkangia replied:
This is already true as this is under `<div class="form-links-container" data-bind=" visible: showFormLinkUI>`

snopoke reviewed (commented):
Rather than calling this `backup` can we call it `fallback`

mkangia committed:
reverse the order of fallback option and put in on top

The stack seems like getting parsed in reverse order on the phone. Hence the top of the stack is checked last which is what we need for fallback option

mkangia committed:
extract stack_static_stack_frames

mkangia committed:
backup -> fallback

mkangia committed:
fix value name

snopoke reviewed (commented):
Looking good but needs some tests.

mkangia committed:
add changes for V2

mkangia committed:
Merge branch 'master' into mk-app-builder-changes-137

mkangia committed:
add v2 diffs

dimagimon commented on corehq/apps/app_manager/suite_xml/post_process/workflow.py:
```diff
@@ -305,7 +309,14 @@ def frame_children_for_module(module_, include_user_selections=True):
+                # for the fallback negative all if conditions/xpath expressions and use that as the xpath for this
+                link_xpaths = [link.xpath for link in form.form_links]
+                if link_xpaths:
+                    negate_of_all_link_paths = ' and '.join(['not(' + link_xpath + ')' for link_xpath in link_xpaths])
```
E501 line too long (118 > 115 characters)

mkangia committed:
add tests for form link fallback

dimagimon commented:
There are 5 errors:

* corehq/apps/app_manager/tests/test_form_workflow.py, line 162 - E123 closing bracket does not match indentation of opening bracket's line
* corehq/apps/app_manager/tests/test_form_workflow.py, line 167 - E123 closing bracket does not match indentation of opening bracket's line
* corehq/apps/app_manager/tests/test_form_workflow.py, line 171 - E501 line too long (127 > 115 characters)
* corehq/apps/app_manager/tests/test_form_workflow.py, line 174 - E501 line too long (125 > 115 characters)
* corehq/apps/app_manager/tests/test_form_workflow.py, line 177 - E501 line too long (123 > 115 characters)

mkangia committed:
dimagimon complaints

emord committed:
Merge branch 'master' into mk-app-builder-changes-137

mkangia committed:
use trans in view, shadows stack_frames from outside the function definition

mkangia committed:
return only the object to be appended instead of passing around the list

mkangia committed:
update v2 view

mkangia committed:
update app manager v2 diffs

mkangia commented:
@snopoke changes done. Skipping lint errors due to incorrect line length check,

snopoke reviewed (approved):
feel free to merge once conflicts are resolved

mkangia committed:
merge master

mkangia committed:
regenerate app manager v2 diffs after resolving conflicts

snopoke commented:
:+1: once build passes
